protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArrayForkChoice.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.protoarray;
+
+import static java.lang.Math.addExact;
+import static java.lang.Math.subtractExact;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.state.Checkpoint;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class ProtoArrayForkChoice {
+  private final ReadWriteLock protoArrayLock = new ReentrantReadWriteLock();
+  private final ReadWriteLock votesLock = new ReentrantReadWriteLock();
+  private final ReadWriteLock balancesLock = new ReentrantReadWriteLock();
```

Comment:
I definitely thought of doing this but then went for the simple locks I have now. I think it's a good idea to get this in this PR.

Reply:
I'm thinking of coming back to this after I'm done with all the tests because I think implementing this might be a little problematic. One potential issue I see is that, with the encapsulation above, we would have to keep the read and write locks specific to functions. Also, I made an exercise of trying to use this API for findHead method in ForkChoiceProtoArray, and I was unable to get it to work unless I followed a pattern similar to 
```
protoArray.withWriteLock(protoArrayWrite -> {
            votes.withWriteLock(votesWrite -> {
                 balances.withWriteLock(balancesWrite -> {
List<Long> deltas = computeDeltas(protoArrayWrite.getIndices(), votesWrite, balancesWrite, newBalances);
```